internal/encryption_backend/vault.go

```diff
@@ -0,0 +1,148 @@
+	ciphertextString := string(ciphertext)
+
+	decryptRequest := schema.TransitDecryptRequest{
+		Ciphertext: ciphertextString,
+	}
+	decryptedResponse, err := v.client.Secrets.TransitDecrypt(context.Background(), v.keyName, decryptRequest, vault.WithMountPath(v.mount))
+	if err != nil {
+		return nil, err
+	}
+	decodedPlaintext, err := base64.StdEncoding.DecodeString(decryptedResponse.Data["plaintext"].(string))
+	if err != nil {
+		return nil, err
+	}
+	return decodedPlaintext, nil
+}
+
+// Encrypt implements SecretBackend.
+func (v VaultBackend) Encrypt(plaintext []byte) ([]byte, error) {
+	encodedPlaintext := base64.StdEncoding.EncodeToString(plaintext)
+
+	encryptRequest := schema.TransitEncryptRequest{
+		Plaintext:  encodedPlaintext,
+		Type:       encryptionType,
+		KeyVersion: 0, // use the latest version of the key
+	}
+	encryptedResponse, err := v.client.Secrets.TransitEncrypt(context.Background(), v.keyName, encryptRequest, vault.WithMountPath(v.mount))
+	if err != nil {
+		return nil, err
+	}
+	return []byte(encryptedResponse.Data["ciphertext"].(string)), nil
+}
+
+// NewVaultBackendWithToken creates a new VaultBackend using token authentication.
+func NewVaultBackendWithToken(endpoint, mount, keyName, token, tlsCaCertificate string, tlsSkipVerify bool, logger *zap.Logger) (VaultBackend, error) {
+	logger.Info("Creating Vault backend", zap.String("endpoint", endpoint), zap.String("mount", mount), zap.String("keyName", keyName), zap.Bool("tls_skip_verify", tlsSkipVerify), zap.String("tls_ca_certificate", tlsCaCertificate))
+	client, err := vault.New(
+		vault.WithAddress(endpoint),
+		vault.WithTLS(vault.TLSConfiguration{
+			ServerCertificate:  vault.ServerCertificateEntry{FromFile: tlsCaCertificate},
+			InsecureSkipVerify: tlsSkipVerify,
+		}),
+	)
+	if err != nil {
+		return VaultBackend{}, err
+	}
+	err = client.SetToken(token)
+	if err != nil {
+		return VaultBackend{}, err
+	}
+	backend := VaultBackend{
+		client:  VaultClient{Auth: &client.Auth, Secrets: &client.Secrets},
+		mount:   mount,
+		keyName: keyName,
+		logger:  logger,
+	}
+	return backend, nil
+}
+
+// NewVaultBackendWithAppRole creates a new VaultBackend using AppRole authentication.
+func NewVaultBackendWithAppRole(endpoint, mount, keyName, roleID, roleSecretID, tlsCaCertificate string, tlsSkipVerify bool, logger *zap.Logger) (VaultBackend, error) {
+	client, err := vault.New(
+		vault.WithAddress(endpoint),
+		vault.WithTLS(vault.TLSConfiguration{
+			ServerCertificate:  vault.ServerCertificateEntry{FromFile: tlsCaCertificate},
+			InsecureSkipVerify: tlsSkipVerify,
+		}),
+	)
+	if err != nil {
+		return VaultBackend{}, err
+	}
+
+	resp, err := client.Auth.AppRoleLogin(context.Background(), schema.AppRoleLoginRequest{
+		RoleId:   roleID,
+		SecretId: roleSecretID,
+	})
+	if err != nil {
+		return VaultBackend{}, err
+	}
+	if err := client.SetToken(resp.Auth.ClientToken); err != nil {
+		logger.Fatal("Error setting token", zap.Error(err))
```

Comment:
Are we sure about the `Fatal` error at this level. Shouldn't the caller decide what to do whit this error?

Reply:
Yes, this is probably better as a warning.